PWGHF/D2H/Tasks/taskDplus.cxx

```diff
@@ -307,53 +500,220 @@ struct HfTaskDplus {
+        fillSparseML<true, false>(candidate, ptBhad, flagBHad, cent, occ);
       }
     }
+  }
+
+  // Run analysis for the generated Dplus candidates
+  /// \param mcGenCollisions are the generated MC collisions
+  /// \param mcRecoCollisions are the reconstructed MC collisions
+  /// \param mcGenParticles are the generated MC particle candidates
+  template <bool fillMl, typename Cand>
+  void runMCGenAnalysis(aod::McCollisions const& mcGenCollisions,
+                        McRecoCollisionsCent const& mcRecoCollisions,
+                        const Cand& mcGenParticles)
+  {
     // MC gen.
-    for (const auto& particle : mcParticles) {
-      auto yGen = RecoDecay::y(particle.pVector(), o2::constants::physics::MassDPlus);
-      if ((yCandGenMax >= 0. && std::abs(yGen) > yCandGenMax) || (std::abs(particle.flagMcMatchGen()) != 1 << aod::hf_cand_3prong::DecayType::DplusToPiKPi)) {
-        continue;
+    float cent{-1.};
+    float occ{-1.};
+    float ptGenB{-1.};
+    int flagGenB{-1};
+
+    for (const auto& mcGenCollision : mcGenCollisions) {
+      const auto recoCollsPerGenMcColl = mcRecoCollisions.sliceBy(recoColPerMcCollision, mcGenCollision.globalIndex());
+      const auto mcParticlesPerGenMcColl = mcGenParticles.sliceBy(mcParticlesPerMcCollision, mcGenCollision.globalIndex());
+      if (storeCentrality && centEstimator != CentralityEstimator::None) {
+        cent = getMcGenCollCentrality(recoCollsPerGenMcColl);
       }
-      fillHistoMCGen(particle);
+      if (storeOccupancy && occEstimator != OccupancyEstimator::None) {
+        occ = getMcGenCollOccupancy(recoCollsPerGenMcColl);
+      }
+
+      for (const auto& particle : mcParticlesPerGenMcColl) {
+        ptGenB = -1;
+        flagGenB = -1;
+        auto yGen = RecoDecay::y(particle.pVector(), o2::constants::physics::MassDPlus);
+        if ((yCandGenMax >= 0. && std::abs(yGen) > yCandGenMax) || (std::abs(particle.flagMcMatchGen()) != 1 << aod::hf_cand_3prong::DecayType::DplusToPiKPi)) {
+          continue;
+        }
+        if (particle.originMcGen() == RecoDecay::OriginType::NonPrompt) {
+          auto bHadMother = mcGenParticles.rawIteratorAt(particle.idxBhadMotherPart() - mcGenParticles.offset());
+          flagGenB = getBHadMotherFlag(bHadMother.pdgCode());
+          ptGenB = bHadMother.pt();
+        }
+        fillHistoMCGen(particle);
+        if constexpr (fillMl) {
+          fillSparseMCGen(particle, ptGenB, flagGenB, cent, occ);
+        }
+      }
+    }
+  }
+
+  /// Get the occupancy
+  /// \param collision is the collision with the occupancy information
+  /// \return collision occupancy
+  template <typename Coll>
+  float getOccupancy(Coll const& collision)
+  {
+    float occupancy = -999.;
+    switch (occEstimator) {
+      case 1:
+        occupancy = collision.trackOccupancyInTimeRange();
+        break;
+      case 2:
+        occupancy = collision.ft0cOccupancyInTimeRange();
+        break;
+      default:
```

Comment:
Why don't you use `OccupancyEstimator` values here?

Reply:
Indeed it was my intention, thanks!